File: wildcard.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
# vim:fenc=utf-8
#
# Copyright © 2015 zhao <zhao@kamel-Desktop>
#
# Distributed under terms of the MIT license.


class Solution:
  # @param s, an input string
  # @param p, a pattern string
  # @return a boolean
  def match(self, indexS, indexP):
    if indexS >= self.lenS and indexP >= self.lenP:
      return True
    if indexS >= self.lenS or indexP >= self.lenP:
      return False
    #if (indexS, indexP) in self.cache:
      #return True
    if self.s[indexS] == self.p[indexP]:
      return self.match(indexS + 1, indexP + 1)
    elif self.p[indexP] == '?':
      return self.match(indexS + 1, indexP + 1)
    elif self.p[indexP] == '*':
      if indexP + 1 >= self.lenP:
        return True
      for i in xrange(indexS, self.lenS):
        temp = self.match(i, indexP + 1)
        if temp:
          #self.cache.add((i, indexP + 1))
          return True
      return False
    else:
      return False

  def removeDupliStar(self, s):
    ans = ""
    for item in s:
      if item == '*':
        if ans and ans[-1] == '*':
          continue
      ans += item
    return ans

  def isMatch(self, s, p):
    self.cache = set()
    if not (s or p):
      return True
    if not (s and p):
      return False
    p = self.removeDupliStar(p)
    self.s = s
    self.lenS = len(s)
    self.p = p
    self.lenP = len(p)
    return self.match(0, 0)

print Solution().isMatch("aa", "a")
print Solution().isMatch("babbbbaabababaabbababaababaabbaabababbaaababbababaaaaaabbabaaaabababbabbababbbaaaababbbabbbbbbbbbbaabbb", "b**bb**a**bba*b**a*bbb**aba***babbb*aa****aabb*bbb***a")

File: wildcard.cpp
/*
 * wildmatch.c
 * Copyright (C) 2015 zhao <zhao@kamel-ThinkPad-X201>
 *
 * Distributed under terms of the MIT license.
 */

#include<stdio.h>
#include<memory.h>
#include<string.h>
#include <stdbool.h>

bool dp[2][100000];
int len_s, len_p;

int count(char *p){
    int ans = 0;
    while(*p){
        if(*(p++) == '*'){
            ans++;
        }
    }
    return ans;
}

bool isMatch(char *s, char *p){
    len_s = strlen(s);
    len_p = strlen(p);
    int count_star = count(p);
    if(len_p - count_star > len_s){
        return false;
    }
    int i, j;
    memset(dp, 0, sizeof(dp));
    dp[0][0] = true;
    for(i = 1; i <= len_p; ++i){
        for(j = 0; j <= len_s; ++ j){
            if(!j){
                dp[i%2][j] = dp[(i + 1)%2][j] && (p[i - 1] == '*');
            }
            else{
                if(p[i - 1] == '?'){
                    dp[i % 2][j] = dp[(i + 1) % 2][j - 1];
                }
                else if(p[i - 1] == '*'){
                    // zero
                    dp[i % 2][j] = dp[i % 2][j - 1];
                    //one
                    dp[i % 2][j] |= dp[(i + 1) % 2][j - 1];
                    //more
                    dp[i % 2][j] |= dp[(i + 1) % 2][j];
                }
                else{
                    dp[i % 2][j] = dp[(i + 1) % 2][j - 1] && (s[j - 1] == p[i - 1]);
                }
            }
        }
    }
    return dp[len_p % 2][len_s];
}

int main(){
    char s1[] = "";
    char s2[] = "*";
    printf("%d\n", isMatch(s1, s2));
}

